dxr/plugins/clang/menus.py

```diff
@@ -201,6 +203,33 @@ def _more_menu_items(self, (qualname,)):
 class FunctionRef(_RefWithDefinition):
     """Ref for function definitions or references"""
 
+    def menu_items(self):
+        """Return a jump-to-definition menu item along with whatever others
+        _more_menu_items() returns.
+
+        """
+        path = self.menu_data[0]
+        if path is None:
+            menu = []
+        else:
+            # We get here when the defloc lies inside the source tree, but for
```

Comment:
> Note to self: understand this next.

I hadn't thought this through far enough, sorry.  There are two cases I was worried about, but now I think the `path is None` check is irrelevant to both of them, which is yet another reason to not derive from `_RefWithDefinition` anymore.

The first case is when you have, for example, a header file with a class that declares a method `C::foo()`.  If the header file is in the tree, and there's a ref in a file in the tree (which includes the header), but the actual implementation of `C::foo` is in a file outside the tree in a compilation unit different from the ref file, then when clang compiles the compilation unit that includes the function ref it will list the declaration in the header as the function def location for `foo` (because the actual implementation location is not in that compilation unit, and so the decl is the best it can do).  So the ref type is in the tree and so we show a context menu and a 'Jump to definition' option for that ref, but in actuality the def is not in the tree and so the query for it returns 0 results.

The other issue is if you have a ref to function `foo()` in the tree, and the implementation file for `foo` is in the tree in a compilation unit different than the ref file, but the header declaring `foo` (which is, for this discussion, the header which the ref in question includes to recognize `foo`), is not in the tree, then I was concerned that you should get a "Jump to definition" menu option because the definition does exist in the tree, but in fact since the header file lies outside the tree you don't get a context menu at all for that ref - its type isn't recognized by dxr.  I guess you could still argue that maybe you should, since the def is actually in the tree (at least maybe for functions if not methods), but I'm okay with not arguing that :-P

So you can get 0 results doing 'Jump to definition' this way (i.e. you can show the 'Jump to definition' option when you shouldn't, and I don't think there's any way around that), but I guess you're not really going to miss any definitions you should have found like I was thinking you could.